### PR TITLE
Unifty httpd in blackholes

### PR DESCRIPTION
**Pull request by blt**
### What does this PR do?

This commit unifies the httpd bits of the blackholes, removing a duplicated loop in three places. The type logic here started out general and got gradually more concrete and there's scope to reduce duplication even further but I don't know that it's a pressing issue.

### CODE REVIEW

**Comment by blt** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/DataDog/lading/1187?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1187** <a href="https://app.graphite.dev/github/pr/DataDog/lading/1187?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/DataDog/lading/1187?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1186** <a href="https://app.graphite.dev/github/pr/DataDog/lading/1186?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1182** <a href="https://app.graphite.dev/github/pr/DataDog/lading/1182?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1181** <a href="https://app.graphite.dev/github/pr/DataDog/lading/1181?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: lading/src/blackhole/common.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+pub(crate) async fn run_httpd<SF, S>(
+    addr: SocketAddr,
+    concurrency_limit: usize,
+    shutdown: Watcher,
+    make_service: SF,
+) -> Result<(), Error>
+where
+    // "service factory"
+    SF: Send + Sync + 'static + Clone + Fn() -> S,
+    // The bounds on `S` per
```
Comment (scottopell) [inline lading/src/blackhole/common.rs:29]: helpful comments, thanks

File: lading/src/blackhole/common.rs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+                        error!("Error accepting connection: {e}");
+                        continue;
+                    }
+                };
+
+                let sem = Arc::clone(&sem);
+                let service_factory = make_service.clone();
+
+                join_set.spawn(async move {
+                    debug!("Accepted connection from {addr}");
```
Comment (scottopell) [inline lading/src/blackhole/common.rs:68]: nit: this should move up a few lines to be closer to when the connection was _actually_ accepted. Doing it across an async operation could create confusing log timestamps.

File: lading/src/blackhole/common.rs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    shutdown: Watcher,
+    make_service: SF,
+) -> Result<(), Error>
+where
+    // "service factory"
+    SF: Send + Sync + 'static + Clone + Fn() -> S,
+    // The bounds on `S` per
+    // https://docs.rs/hyper/latest/hyper/service/trait.Service.html and then
+    // made concrete per
+    // https://docs.rs/hyper-util/latest/hyper_util/server/conn/auto/struct.Builder.html#method.serve_connection.
```
Comment (scottopell) [inline lading/src/blackhole/common.rs:32]:
```suggestion
    // https://docs.rs/hyper-util/latest/hyper_util/server/conn/auto/struct.Builder.html#method.serve_connection
```

trailing period breaks the hyperlink

**Review by scottopell — APPROVED** [review]
Change looks good for unifying the common functionality here

File: lading/src/blackhole/common.rs (hunk 3 above)
Comment (blt) [reply lading/src/blackhole/common.rs]: Addressed in [a1e268c](https://github.com/DataDog/lading/pull/1187/commits/a1e268c2f29ecd894ada9f77e5744d7ac68451fa).

File: lading/src/blackhole/common.rs (hunk 2 above)
Comment (blt) [reply lading/src/blackhole/common.rs]: Ah yep, quite right. Addressed in [606b878](https://github.com/DataDog/lading/pull/1187/commits/606b87864a7edbd9ec1acb4cf358c987cf62c4ae).

**Review by GeorgeHahn — APPROVED** [review]
Catching up to the slack thread on this - I think this is functionally identical to the behavior of the previous tower concurrency limit + load shed combination.

**Comment by blt** [comment]
### Merge activity

* **Dec 30, 2:21 PM EST**: A user merged this pull request with [Graphite](https://app.graphite.dev/github/pr/DataDog/lading/1187).

**Pull request merged**